### ISSUE TITLE
What does the mask means in the representation loss? 

### ISSUE THREAD

**Issue by huangtiansheng** [domenicrosati/representation-noising#1]
Hi authors,
Thanks for the amazing work. 

When I read through the code, I am not clear about what the "mask" in the following code mean. It seems that it is used to mask out some overlap token between harmful input and harmless input in the hidden activation. As it seems that this is not mentioned in the paper, I am wondering what is the exact motivation for this procedure? Thanks a lot!   

```
mask = ~torch.eq(harmful_batch['input_ids'], harmless_batch['input_ids'])
noise_loss = 0
for i, hidden in enumerate(harmful_activations):
    hiddens_mask = mask.unsqueeze(-1).expand(hidden.size()).to(hidden.device)
    hiddens = hidden * hiddens_mask
    gaussian = torch.randn_like(hiddens).to(hidden.device) * hiddens_mask
    noise_loss += mmd_loss(hiddens.view(hiddens.size(0), -1), gaussian.view(gaussian.size(0), -1)).to(DEVICE)
```

**Comment by domenicrosati** [comment]
Sorry we are adding the code as fast as possible so some things might not be clear until then.

So the MASK operation is in the paper (Line 6 of Algorithmn 1 - Appendx B.1  Implementation). Where we say:
> We also draw the reader’s attention to the MASK operation. This is because
we use paired refusal data for our method where the first part of a question or prompt is shared
between the harmful and harmless target text. We don’t want to perform gradient ascent or noise
the representations of these shared tokens which is why a mask is required. The mask is simply the
tokens that a paired sample has in common.



I realize this could be clearer! but essentailly what this means is that with the paired refusal data we only compute losses for what is different between the refusals and harmful answers.

Let me know if that has clarified things otherwise I can give some more details.

**Comment by huangtiansheng** [comment]
Thanks for the information. Yeah, both the two small tricks look interesting and probably needs a great amount of effort to uncover. Thank you!
